litestar/controller/generic.py

```diff
@@ -0,0 +1,243 @@
+    get_many_http_method: HttpMethod = HttpMethod.GET
+    """The HTTP method for get many operations."""
+
+    def __init__(self, owner: Router) -> None:
+        """Initialize a controller.
+
+        Should only be called by routers as part of controller registration.
+
+        Args:
+            owner: An instance of :class:`Router <.router.Router>`
+        """
+        super().__init__(owner=owner)
+
+        if not getattr(self, "repository_type", None):
+            raise ImproperlyConfiguredException("generic controllers must define a `repository_type` attribute")
+        # TODO add DTO init logic here
+
+    def _normalize_annotation(self, key: str, annotation: Any) -> Any:
+        if (origin := get_origin(annotation)) and (args := get_args(annotation)):
+            safe_origin = get_safe_generic_origin(origin, origin)
+            return safe_origin[tuple(self._normalize_annotation(key, arg) for arg in args)]
+        if annotation is ModelT:  # type: ignore[misc]
+            return self.model_type
+        if key in ("item_id", "item_ids") and annotation is Any:
+            return self.id_attribute_type
+        return annotation
+
+    def get_route_handlers(self) -> list[BaseRouteHandler]:
+        route_handlers = super().get_route_handlers()
+
+        for method_name in GENERIC_METHOD_NAMES:
+            method = getattr(self, method_name)
+            if not isinstance(method, BaseRouteHandler):
+                handler_path = getattr(self, f"{method_name}_handler_path")
+                operation_id = getattr(self, f"{method_name}_operation_id")
+                http_method = getattr(self, f"{method_name}_http_method")
+
+                # we are replacing the generic parameters with the concrete types in the method `__annotations__`
+                # this is required to ensure we model the signautre correctly, and generate the schemas as required.
+                for k, v in get_type_hints(method, localns={"Request": Request}).items():
+                    method.__annotations__[k] = self._normalize_annotation(k, v)
+
+                route_handler = HTTPRouteHandler(
+                    handler_path.replace("path_param_type", self.path_param_type),
+                    http_method=http_method,
+                    operation_id=operation_id.replace("{model_name}", self.model_type.__name__),
+                    signature_namespace={
+                        self.model_type.__name__: self.model_type,
+                        self.id_attribute_type.__name__: self.id_attribute_type,
+                    },
+                )(method)
+                route_handler.owner = self
+                route_handlers.append(route_handler)
+
+        route_handlers.sort(key=attrgetter("handler_id"))
+        return route_handlers
+
+    @classmethod
+    def get_generic_annotations(cls) -> tuple[type, type]:
+        try:
+            if (generic_bases := getattr(cls, "__orig_bases__", None)) and (
```

Comment:
`__orig_bases__` generally contain `tuple[Any, ...]`, not `tuple[type, ...]`, because there are multiple special forms that are allowed.

Reply:
See https://github.com/python/typeshed/blob/703ed36d5a5c9505c903ea2182e6eed679d9bddb/stdlib/types.pyi#L586 and https://github.com/python/typeshed/issues/7811

Reply:
Cool,  addressed